IP address parsing refactoring

jukkar opened the pull request:
As there is more uses for IP address parsing, create a generic net_ipaddr_parse() function for it and replace current functionality in DNS resolver by it. Update the unit tests to test the parsing function.

jukkar committed:
net: Add function to parse IP address string

The net_ipaddr_parse() will take a string with optional port
number and convert its information into struct sockaddr.
The format of the IP string can be:
     192.0.2.1:80
     192.0.2.42
     [2001:db8::1]:8080
     [2001:db8::2]
     2001:db::42

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar committed:
net: tests: Add tests for IP address parsing

Add IPv4 and IPv6 address parsing tests.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar committed:
net: dns: Use IP address parsing function in init

This commit removes IP address parsing from DNS init and
replaces it by call to net_ipaddr_parse().

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

pfalcon commented:
My understanding is that this function is intended for usage convenience first of all, then why:

> @param str_len Length of the string to be parsed.

Why not just take null-terminated string?

Especially that there's code like:

> ptr = strstr(str, "]");

This won't work properly with given length (e.g. if we want to parse address is a larger buffer, it may find "]" outside the string we want to parse).

pfalcon commented on subsys/net/ip/utils.c:
```diff
@@ -518,3 +518,222 @@ bool net_header_fits(struct net_pkt *pkt, u8_t *hdr, size_t hdr_size)
+		if ((ptr + 1) < (str + str_len) && *(ptr + 1) == ':') {
+			memcpy(ipaddr, ptr + 2, str_len - end);
+			ipaddr[str_len - end] = '\0';
+			port = strtol(ipaddr, &endptr, 10);
```
Couldn't this port parsing be rewritten somehow to be common for both IPv6 & IPv4, instead of being duplicated?

jukkar replied:
Perhaps, I did not see too much reason to create a function for this as it is just couple of lines.

pfalcon commented on include/net/net_ip.h:
```diff
@@ -870,6 +870,29 @@ char *net_addr_ntop(sa_family_t family, const void *src,
+ * and optional port, and store the information in user supplied
+ * sockaddr struct.
+ *
+ * @details Syntax of the IP address string:
```
I'd say this should specify explicitly what will be on sockaddr if port is not specified (e.g. port field set to 0, or yet better not touched, to let the caller to initialize it to default value prior to call).

jukkar replied:
Sure, that makes sense.

pfalcon commented:
@jukkar: Also, as DNS was mentioned, pls have a look at https://jira.zephyrproject.org/browse/ZEP-2562 when you get a chance (implementing that definitely goes to 1.10).

jukkar commented:
> Why not just take null-terminated string?

If we want to parse address from middle of a string. This way we do not need to copy to temporary buffer and parse that.

jukkar commented:
> This won't work properly with given length (e.g. if we want to parse address is a larger buffer, it may find "]" outside the string we want to parse).

The lengths are checked later in the function. I have unit tests that should check this.

jukkar commented:
> The lengths are checked later in the function. I have unit tests that should check this.

After checking the unit tests, I did not actually test this condition. I will send a new version that addresses this issue.

jukkar commented:
> Also, as DNS was mentioned, pls have a look at https://jira.zephyrproject.org/browse/ZEP-2562 when you get a chance (implementing that definitely goes to 1.10).

As I commented in bug, we should fix the DNS resolver. One simple solution we could use, is to call this new net_ipaddr_parse() in DNS resolver, and if that fails, then try to resolve from DNS server.

jukkar committed:
net: dns: Do not resolve numeric IP address

If the query name is already in numeric format, there is no
need to send the query string to DNS server as we can just
convert it ourselves.

Jira: ZEP-2562

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar commented:
@pfalcon Add new patch to this set that fixes the numeric DNS query issue you reported.

pfalcon reviewed (dismissed):
Thanks for the changes and adding numeric addresses support to DNS query! This fails CI so, but let me proactively approve this, as I'm going on vacation the next day.

lpereira commented on subsys/net/ip/utils.c:
```diff
@@ -518,3 +518,230 @@ bool net_header_fits(struct net_pkt *pkt, u8_t *hdr, size_t hdr_size)
+	char *endptr;
+	int tmp;
+
+	tmp = strtol(buf, &endptr, 10);
```
`errno` has to be set to `0` before calling `strtol()`.  Might be a good idea to use `strtoul()` instead since it's unsigned anyway.

jukkar replied:
It looks like we do not need errno when checking the conversion failures if we use endptr as suggested in strtoul() manual page.

lpereira commented on subsys/net/ip/utils.c:
```diff
@@ -518,3 +518,230 @@ bool net_header_fits(struct net_pkt *pkt, u8_t *hdr, size_t hdr_size)
+static bool convert_port(const char *buf, u16_t *port)
+{
+	char *endptr;
+	int tmp;
```
While `int` and `long` are the same size, it's better to use the same return type for `strtol()` here: `long int`.

lpereira commented on subsys/net/ip/utils.c:
```diff
@@ -518,3 +518,230 @@ bool net_header_fits(struct net_pkt *pkt, u8_t *hdr, size_t hdr_size)
+	tmp = strtol(buf, &endptr, 10);
+	if ((endptr == buf && tmp == 0) ||
+	    !(*buf != '\0' && *endptr == '\0') ||
+	    tmp > 0xffff || tmp < 0) {
```
If using `strtoul()` instead, you might change the range check with a cast: `if ((unsigned long)(unsigned short)tmp != tmp)`. This often generates tighter code.

jukkar replied:
Sure

lpereira commented on subsys/net/ip/utils.c:
```diff
@@ -518,3 +518,230 @@ bool net_header_fits(struct net_pkt *pkt, u8_t *hdr, size_t hdr_size)
+		char ipaddr[INET6_ADDRSTRLEN + 1];
+		int end, len;
+
+		ptr = strstr(str, "]");
```
Please use `memchr()` here: since there's a `str_len` parameter, a user might pass a non-zero-delimited string.  Moreover, `strstr()` is overkill to scan for a single character.

jukkar replied:
Indeed, we can also then remove the second ptr check in next line.

lpereira commented on subsys/net/ip/utils.c:
```diff
@@ -518,3 +518,230 @@ bool net_header_fits(struct net_pkt *pkt, u8_t *hdr, size_t hdr_size)
+	}
+
+	count = i = 0;
+	while (str[i]) {
```
Nitpick: Why not a `for (count = i = 0; str[i]; i++) ...` instead?

jukkar replied:
Yes, and "for (count = i = 0; str[i] && i < str_len; i++) {" is even better as it checks the string length too.

lpereira commented on subsys/net/ip/utils.c:
```diff
@@ -518,3 +518,230 @@ bool net_header_fits(struct net_pkt *pkt, u8_t *hdr, size_t hdr_size)
+
+			net_sin6(addr)->sin6_port = htons(port);
+
+			NET_DBG("IPv6 server %s port %d",
```
Are these always servers? Wouldn't `host` be better?

jukkar replied:
Left over when the code was in DNS resolve library, I will convert to "host" instead.

jukkar commented:
New version which adds more unit tests and updates the code according to comments.

jukkar commented:
@lpereira could you review / re-approve the changes?

lpereira commented on subsys/net/ip/utils.c:
```diff
@@ -518,3 +518,230 @@ bool net_header_fits(struct net_pkt *pkt, u8_t *hdr, size_t hdr_size)
+	int end, len;
+
+	len = min(NET_IPV4_ADDR_LEN, str_len);
+	end = min(len, strlen(str));
```
This passed from the other review, but `strnlen()` should be used here, since `str` is not guaranteed to have a NUL terminator.  Other uses of `strlen()` here should be changed.  Apparently, minimal libc does not have a `strnlen()` implementation, but it's trivial to add one.

lpereira commented on subsys/net/ip/utils.c:
```diff
@@ -518,3 +518,230 @@ bool net_header_fits(struct net_pkt *pkt, u8_t *hdr, size_t hdr_size)
+	char ipaddr[NET_IPV4_ADDR_LEN + 1];
+	int end, len;
+
+	len = min(NET_IPV4_ADDR_LEN, str_len);
```
Do we have `alloca()` support?  If so, these things copying `str` into `ipaddr` could become just a call to `strndupa()`.   [Soletta](https://github.com/solettaproject/soletta/blob/master/src/lib/common/sol-missing.h#L30-L38) has a simple implementation that you can borrow (same license, Intel owns the copyright).

jukkar replied:
Lets not introduce yet another dependecy here :)

lpereira reviewed (changes requested):
Could `net_ipaddr_parse()` be either split into smaller functions, with one per configuration variant, or at least split between `parse_ipv4` and `parse_ipv6` functions? The amount of copy-paste in parsing code worries me a bit.

jukkar commented:
> Could net_ipaddr_parse() be either split into smaller functions, with one per configuration variant, or at least split between parse_ipv4 and parse_ipv6 functions? The amount of copy-paste in parsing code worries me a bit.

Yes, that makes sense. I sent a new version that does that. Also added more test cases to unit tests.
